Adding sub dialogue for formatting of summary tables

Ivanluv opened the pull request:
fixes #6770
This is still in progress

lloyddewit commented:
@Ivanluv thanks for taking on this issue.
If a PR is not yet ready for review, then you can set it to draft (see the 'convert to draft' link in top-right of the screen).
![image](https://user-images.githubusercontent.com/57253949/142996854-96f974cb-1652-4f3e-a9f5-6cd54f938b1f.png)

Ivanluv commented:
> @Ivanluv thanks for taking on this issue. If a PR is not yet ready for review, then you can set it to draft (see the 'convert to draft' link in top-right of the screen). ![image](https://user-images.githubusercontent.com/57253949/142996854-96f974cb-1652-4f3e-a9f5-6cd54f938b1f.png)

Thanks for this

lloyddewit commented:
@Ivanluv Please could you clarify the status of this PR? thanks

Ivanluv commented:
@africanmathsinitiative/developers  this is ready for review. @lilyclements  could you take a look too

lloyddewit commented:
@Ivanluv Thank you! this looks like a significant PR.
@rdstern Please could you do an initial test?
@N-thony Please could you do an initial review?
Thanks

N-thony reviewed (changes requested):
Could you also add at the top one line space before the import?

' R- Instat
' Copyright (C) 2015-2017
'
' This program is free software: you can redistribute it and/or modify
' it under the terms of the GNU General Public License as published by
' the Free Software Foundation, either version 3 of the License, or
' (at your option) any later version.
'
' This program is distributed in the hope that it will be useful,
' but WITHOUT ANY WARRANTY; without even the implied warranty of
' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
' GNU General Public License for more details.
'
' You should have received a copy of the GNU General Public License 
' along with this program.  If not, see <http://www.gnu.org/licenses/>.

N-thony commented on instat/sdgFormatSummaryTables.vb:
```diff
@@ -1,3 +1,552 @@
+        clsTabOptionsFunction, clsBorderWeightPxFunction, clsFootnoteSubtitleLocationFunction, clsTabFootnoteSubtitleFunction, clsFootnoteCellBodyFunction,
+        clsStyleListFunction, clsStubHeadFunction As New RFunction
+    Private clsPipeOperator, clsMutableOperator, clsJoiningOperator As New ROperator
+    Private bControlsInitialised = False
```
As Boolean?

Ivanluv replied:
Thank you for noticing this.Done

rdstern commented:
As @lloyddewit has said, this is a major new feature.  This sub-dialogue is starting (I assume) a similar "system" to the sub-dialogue for the graphs.
a) So please could it also be added to the Frequency Tables dialogue at the same time.  I assume it will be identical there, and this will help in testing the features.  I assume it will replace the disabled Options button?
b) So I note first a few small items.  I would like to merge it as soon as possible, but assume improvements will be on-going work for a bit of time yet.
c) On the title tab it gives me an error when I tick anything, but leave it all empty.  That's for any of the 3 categories.  Could it ignore the tick if it is then empty? Or capture that in a more friendly way?  It doesn't seem to need everything completed, but does need something.
d) The footnote control is quite small.  There is space to make them larger without changing the size of the sub-dialogue.
e) Add sourcenote seems a nice command name, but perhaps could be Add Source on the dialogue.
f) For the footnote I see we can only have a single footnote?  How could we have (say) 2 footnotes?
g) On the main dialogue, it needs to become a bit shorter.  I wonder about:
1) Moving the Summary button to above the Summary control.
2) Making the Summary multiple receiver a bit smaller.
3) Move the Store Output to the bottom - where the summaries button is now?
4) Move the Margin Name up a git, so on the same level as the margin checkbox
5) Then the 3 margin options can go across.
h) Will there be more options on the titles tab, for example font, sixe?
i) The header is either for rows or columns?  Should there be 2 tables, so each can be specified?
j) Most are set to NULL as the default, but the 2 colour ones are set initially to a (confusing) number - though NULL is available.
k) On the Table tab I am not sure what to expect.  Can Background Colour be without the capital G.  I Tried to make it yellow and it coloured the title and footnote.  Is that what it was meant to do?  Can I get the sort of formatting colours that Excel gives, with alternate rows, etc?

Ivanluv commented:
@rdstern  could you have a look again
These are the changes I have added:

1. Added the format table sub dialogue to the frequency table
2. Fixed the error On the title tab when  anything is ticked  while leave it all empty. This I have fixed by  having the footnote ,title and  source note functions only when the labels are provided
3. Made design changes to the summary dialogue and minor changes to the sub dialogue
4. Added the option for the second footnote
5. Added the font and size options for the title

> 5\. j) Most are set to NULL as the default, but the 2 colour ones are set initially to a (confusing) number - though NULL is available

Those numbers are the R defaults values for those colours



> 5\. k) On the Table tab I am not sure what to expect.  Can Background Colour be without the capital G.  I Tried to make it yellow and it coloured the title and footnote.  Is that what it was meant to do?  Can I get the sort of formatting colours that Excel gives, with alternate rows, etc?

I might need help with these

Ivanluv commented on instat/sdgFormatSummaryTables.vb:
```diff
@@ -1,3 +1,552 @@
+        ucrInputStyleTextSize.SetRDefault("NULL")
+        ucrInputStyleTextSize.SetLinkedDisplayControl(lblStyleTextSize)
+
+
```
Done

rdstern commented:
@Ivanluv and possibly @lilyclements I have a curious feature in the new summary tables - which seems to be coming on fine.  And I am excited by my latest discovery - so let me start with that!
a) I mentioned to Ivan recently that I am keen for the Describe > One Variable > Summarise dialogue to be improved.  I have also found an issue with that dialogue when I ask for proportions.  It gives missing values and I seem to remember that this also happened - and was corrected - in this summary dialogue.
b) Then I found one of the items I wanted what to be able to change the orientation of the summaries in the one variable dialogue.  And of course, this is possible already in this summary dialogue!
c) I think you can see where this is all heading and (for me) it is simple and exciting!  The one variable dialogue (Customised) IS the summary dialogue, made simple - and with no factors!!!

So, back to the summary dialogue itself.  After that exciting discovery I used the summary dialogue with the recent R gui comparison data table that has been circulated.  Read 43 rows, so not the total at the bottom.  Then use your new summary dialogue.  I get all the results I want, but with 2 oddities!  Here is the dialogue and the output:

![image](https://user-images.githubusercontent.com/11226469/148585998-73ed01df-932e-4f98-a039-00293df3d928.png)

Here are the results:
![image](https://user-images.githubusercontent.com/11226469/148586434-8953a6e2-4577-47f6-9658-30b4daf3146c.png)

However, they are not true!  It is like the Morecambe and Wise thing = "all the numbers are there, but not necessarily in the right order! " Both sets of names seem to be alphabetical, but it has really shuffled the columns I think!

The default is correct, (untick, both summaries as a factor and displays as rows) but then everything is in one very long row.  I think all other combinations add a shuffle!

This is a bit urgent to fix.  I should check it in 7.3.1!

rdstern reviewed (changes requested):
@Ivanluv thanks for all those changes.  Just a couple of cosmetic changes.
a) It now says Display Outer and that should go back to Display Outer Margins.  Then you could delete the Margin from Margin Name, so there is space.
b) You have shortened the Factors and Variates controls.  I was thinking of leaving those and shortening the Variables control instead.  I think that will still be ok with the space.
c) And perhaps make the group box for the margins a little less generous, so it is still just a bit smaller overall?
d) Then there is the error we found to fix.  
e) then I hope @lloyddewit or @shadrackkibet can check the code soon, so we can merge.

Ivanluv commented:
Am closing this PR due to the many conflicts
